### DISCUSSION TITLE
Using Vercel (or a similar tool) to deploy previews of changes

### DISCUSSION THREAD

**Discussion by Ein-Tim** [corona-warn-app/cwa-website#3287] [Development] [answered]
## Feature description

[Vercel](https://vercel.com) offers a [GitHub App](https://github.com/apps/vercel) which deploys changes to a test environment so that they can there be reviewed and tested. It would be great if the CWA could use this feature, either from Vercel or build something like this themselves. 

## Problem and motivation
1. PR reviewing goes way easier for anyone, team and community
2. The server where the change is deployed to for review can be configured exactly as the PROD environment to be able to detect all issues before going live. 

## Is this something you're interested in working on

I can support, but I can't build something like this myself. Here is the Vercel documentation on this: https://vercel.com/docs/concepts/git/vercel-for-github

---
Internal Tracking ID: [EXPOSUREAPP-13074](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-13074)

**Comment by svengabr** [comment]
We already had a similar discussion internally a while back before we introduced the new FAQ.
The outcome was that the costs were too high in comparison to the benefit it would give.

I totally agree that it would be nice to have. We can for sure bring it up in our meetings.

**Reply by MikeMcC399** [reply]
What is the remaining life-time for this site? It would only make sense if the website will continue for a longer time and there are big changes expected.

**Comment by Ein-Tim** [comment]
@svengabr

Thanks for the feedback! Good to know that you discussed it already!

Yes, absolutely, setting up this is a huge workload, but it makes day-to-day routine work so much easier, that it could very well safe some work hours, once implemented.

It would be great if you could re-discuss this, maybe also mirror it to JIRA.

Thank you!

**Comment by Ein-Tim** [comment]
Well, if you're not doing it, I do it, for example here: https://github.com/corona-warn-app/cwa-website/pull/3344.

This took me literally 10 minutes to set up, and it is **sooo** worth it as it works automatically, and I don't have to deal with any commands anymore! I guess the only obstacle would be to get this approved by stakeholders, but I will see what I can do about that!

**Comment by MikeMcC399** [comment, marked as answer]
@Ein-Tim 

I'm interested to see your preview using Vercel which is displaying a public preview of the static website, that is otherwise only available privately from `npm start` and on your `http://localhost:8000/en/`.

For individual community members, I'm not sure if this can be used like you are using it, since the free [Hobby plan](https://vercel.com/pricing) says that it is for personal or non-commercial projects. Or are you paying for a Pro subscription ($20 / month)? I guess you would have to check with Vercel if they consider CWA to be a non-commercial project or not. The CWA app use is free of charge, however the development is commercially funded by the German government, as you well know. 

If it were to be an integral part of the workflow for CWA, then probably Vercel [Pro or Enterprise subscriptions](https://vercel.com/pricing) would be necessary. Did you look at how this could be technically integrated into the general workflow for team and community members? 

I can't really see this happening as a project investment if the next phase is the ramp down. It's a nice idea though.

**Comment by Ein-Tim** [comment]
@MikeMcC399 

> For individual community members, I'm not sure if this can be used like you are using it, since the free Hobby plan, says that it is for personal or non-commercial projects. Or are you paying for a Pro subscription ($20 / month)

No, I'm not paying for a Pro subscription. I set Vercel up for another, unrelated, repository and thought I just give it a try if it can build my for of the cwa-website. I would consider my work on the CWA, at the moment, to be my hobby, so it's a personal matter, and it's also a personal use of Vercel.
But I guess one could argue over this reasoning.

> If it were to be an integral part of the workflow for CWA, then probably Vercel Pro or Enterprise subscriptions would be necessary. Did you look at how this could be technically integrated into the general workflow for team and community members?

Yeah, right, for official use on this repository a Pro/Enterprise subscription would certainly be necessary. There are, at least it feels like this, endless possibilities to configure Vercel, starting from only deploying previews when manually requested, over integrating it into a GitHub actions workflow, to automatically deploy previews for any commits pushed to any branch of this repository.

> I can't really see this happening as a project investment if the next phase is the ramp down. It's a nice idea though.

You're certainly right that this is unlikely to be implemented in the current phase the project is in!
But in general, maybe also as a learning for future projects, automated previews of changes are defined worth looking into!

I will mark  this discussion as answered by @MikeMcC399 now, as I believe it is as he said; we won't see this anymore.

**Reply by brianebeling** [reply]
Thank you for the suggestion TIm and bringing the Vercel topic up again. Rest assured that we are not actively ignoring your suggestions, After all many of us working on this project are not only working on it because we stand behind the Corona-Warn-App, or because this is our job, but also for the interest and passion in tech. Part of that is also experiencing new tooling and trying to improve the process.

We did discuss the usage of Vercel or a similar tool very early when the idea initially came up, but could not come to a satisfying conclusion then already. As Mike pointed out the project is commercially funded and and as with any idea or potential improvement, multiple stakeholders are involved in a decision like this. And then there was also the concern of the project potentially coming to an end soon.

It won't be implemented for this project, since the ramp down is close - however - we made a note to look at this and other tools for future projects. Thanks again for your input, you aren't wrong that this specific preview feature would have made a lot of things easier and save a bit of time here and there.
